### PR TITLE
Add log viewer

### PR DESCRIPTION
**Pull request by Zarel**
I appreciate Mia for taking a stab at this in #6441, but in hindsight there are enough things I want done a specific way that I can only do this myself.

### CODE REVIEW

**Comment by Zarel** [comment]
Shit, this was supposed to be a draft.

File: server/chat-plugins/logs.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+			return LogViewer.error("SPL team discussions are super secret.");
+		}
+		if (roomid.startsWith('wcop') && !user.can('rangeban')) {
+			return LogViewer.error("WCOP team discussions are super secret.");
+		}
+		if (room) {
+			if (!room.checkModjoin(user) && !user.can('bypassall')) {
+				return LogViewer.error("Access denied");
+			}
+			if (!this.can('lock', null, room as BasicChatRoom)) return;
```
[the remark this answers is not in the document]
Comment (Zarel) [reply server/chat-plugins/logs.ts]: Actually yes.

**Comment by Zarel** [comment]
> Also, please abstract out the reader / access logs to handle plugging in other storage impls.

Mia may have fell for that, but I won't. Abstract it out yourself. :p

**Comment by Zarel** [comment]
`LogReader` and `LogReaderRoom` are classes, that's already plenty abstract imo.

**Comment by TheJetOU** [comment]
> `LogReader` and `LogReaderRoom` are classes, that's already plenty abstract imo.

Yes, you were halfway there. You just needed to make a `LogReader` interface, create an `FS` impl for it, then refactor `LogReader` to take in a `LogReader` impl.

**Comment by Zarel** [comment]
Yeah, see, that's completely unnecessary. `LogReader` is already an interface; TS auto-generates interfaces from classes; you can write `LogReaderDB implements LogReader`. There's no need to noisy-up my code for this. :[

**Pull request merged**